addons/website_sale_digital/controllers/main.py

```diff
@@ -65,42 +47,44 @@ def orders_followup(self, order=None, **post):
     def download_attachment(self, attachment_id):
         # Check if this is a valid attachment id
         attachment = request.env['ir.attachment'].sudo().search_read(
-            [('id', '=', int(attachment_id))],
-            ["name", "datas", "file_type", "res_model", "res_id", "type", "url"]
-        )
+            [('id', '=', int(attachment_id)), ('product_downloadable', '=', True)], ["name", "download_count", "datas_fname", "datas", "res_model", "res_id", "type", "url"])
 
         if attachment:
             attachment = attachment[0]
         else:
             return redirect(self.orders_page)
 
         # Check if the user has bought the associated product
+
         res_model = attachment['res_model']
         res_id = attachment['res_id']
-        purchased_products = request.env['account.invoice.line'].get_digital_purchases()
-
-        if res_model == 'product.product':
-            if res_id not in purchased_products:
-                return redirect(self.orders_page)
-
-        # Also check for attachments in the product templates
-        elif res_model == 'product.template':
-            P = request.env['product.product']
-            template_ids = map(lambda x: P.browse(x).product_tmpl_id.id, purchased_products)
-            if res_id not in template_ids:
-                return redirect(self.orders_page)
-
-        else:
+        partner = request.env.user.partner_id
+        purchased_products = request.env['sale.order.line'].sudo().search(
+            ['&', ('order_id.partner_id', '=', partner.id),
+                  '|', ('order_id.payment_tx_id.state', '=', "done"),
```

Comment:
This will be highly inefficient. It will be replaced by all ids of paid sale orders (order_id, in, [Array]). On a real DB I fear this will be slow.